### PR TITLE
CPM-926: Add Sequence

### PR DESCRIPTION
**Pull request by pierallard**
C.f. JIRA card

### CODE REVIEW

File: components/identifier-generator/back/src/Infrastructure/Symfony/Install/InstallerSubscriber.php [hunk 1]
```diff
@@ -53,5 +53,17 @@ public function updateSchema(): void
...
+                `attribute_id` INT NOT NULL,
+                `identifier_generator_uuid` binary(16) NOT NULL,
+                `prefix` VARCHAR(255) NOT NULL,
+                `last_allocated_number` BIGINT UNSIGNED NOT NULL,
+                CONSTRAINT `FK_SEQ_ATTRIBUTEID` FOREIGN KEY (`attribute_id`) REFERENCES `pim_catalog_attribute` (`id`) ON DELETE CASCADE,
+                CONSTRAINT `FK_SEQ_IDENTIFIERGENERATORUUID` FOREIGN KEY (`identifier_generator_uuid`) REFERENCES `pim_catalog_identifier_generator` (`uuid`) ON DELETE CASCADE,
+                INDEX index_identifier_generator_sequence (`attribute_id`, `identifier_generator_uuid`, `prefix`, `last_allocated_number`)
+            ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
+            SQL);
+        // TODO: should we add a unique index on `attribute_id`, `identifier_generator_uuid`, `prefix` ?
```
Comment (pierallard) [inline components/identifier-generator/back/src/Infrastructure/Symfony/Install/InstallerSubscriber.php:67]: TODO

File: components/identifier-generator/back/tests/EndToEnd/Infrastructure/Subscriber/SetIdentifiersSubscriberEndToEnd.php [hunk 2]
```diff
@@ -55,6 +55,19 @@ public function it_should_generate_an_identifier_on_create(): void
         Assert::assertSame('AKN-050', $productFromDatabase->getValue('sku')->getData());
     }
 
+    /** @test */
+    public function it_should_generate_several_identifiers_on_create(): void
+    {
+        $this->createIdentifierGenerator();
+        $productsFromDatabase = $this->createProducts(5);
+        ;
```
Comment (pierallard) [inline components/identifier-generator/back/tests/EndToEnd/Infrastructure/Subscriber/SetIdentifiersSubscriberEndToEnd.php:63]: ;

File: components/identifier-generator/back/src/Infrastructure/Symfony/Install/InstallerSubscriber.php [hunk 3]
```diff
@@ -53,5 +53,18 @@ public function updateSchema(): void
...
+                `identifier_generator_uuid` binary(16) NOT NULL,
+                `prefix` VARCHAR(255) NOT NULL,
+                `last_allocated_number` BIGINT UNSIGNED NOT NULL,
+                CONSTRAINT `FK_SEQ_ATTRIBUTEID` FOREIGN KEY (`attribute_id`) REFERENCES `pim_catalog_attribute` (`id`) ON DELETE CASCADE,
+                CONSTRAINT `FK_SEQ_IDENTIFIERGENERATORUUID` FOREIGN KEY (`identifier_generator_uuid`) REFERENCES `pim_catalog_identifier_generator` (`uuid`) ON DELETE CASCADE,
+                UNIQUE INDEX sequence_attribute_identifier_prefix (attribute_id, identifier_generator_uuid, prefix),
+                INDEX index_identifier_generator_sequence (`attribute_id`, `identifier_generator_uuid`, `prefix`, `last_allocated_number`)
+            ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci
+            SQL);
+        // TODO: should we add a unique index on `attribute_id`, `identifier_generator_uuid`, `prefix` ?
```
Comment (Beninho) [inline components/identifier-generator/back/src/Infrastructure/Symfony/Install/InstallerSubscriber.php:68]: TODO is still here be it is already resolved

**Review by Beninho — APPROVED** [review]

**Review by mmetayer — APPROVED** [review]

File: components/identifier-generator/back/src/Application/Generate/GenerateIdentifierCommand.php [hunk 4]
```diff
@@ -25,21 +22,8 @@ public static function fromIdentifierGenerator(IdentifierGenerator $identifierGe
...
-     */
-    public function getProperties(): array
-    {
-        return $this->identifierGenerator->structure()->getProperties();
-    }
-
-    public function getTarget(): Target
-    {
-        return $this->identifierGenerator->target();
+        return $this->identifierGenerator;
```
[the remark this answers is not in the document]
Comment (pierallard) [reply components/identifier-generator/back/src/Application/Generate/GenerateIdentifierCommand.php:27]: https://akeneo.atlassian.net/browse/CPM-928

File: components/identifier-generator/back/src/Application/Generate/GenerateIdentifierHandler.php [hunk 5]
```diff
@@ -33,31 +33,32 @@ public function __construct(
...
-        $properties = $command->getProperties();
-        $target = $command->getTarget();
+        $identifierGenerator = $command->getIdentifierGenerator();
 
         $result = '';
-        foreach ($properties as $property) {
+        foreach ($identifierGenerator->structure()->getProperties() as $property) {
             if ($result !== '') {
-                $result .= $delimiter?->asString() ?? '';
+                $result .= $identifierGenerator->delimiter()?->asString() ?? '';
```
[the remark this answers is not in the document]
Comment (pierallard) [reply components/identifier-generator/back/src/Application/Generate/GenerateIdentifierHandler.php:41]: https://akeneo.atlassian.net/browse/CPM-929

**Pull request merged**